### ISSUE TITLE
use mobilenetv3 may Get better results?

### ISSUE THREAD

**Issue by 121786404** [biubug6/Pytorch_Retinaface#3]

**Comment by biubug6** [comment]
You're right. The more expressive the backbone network is, the better it performs, but the slower the speed will be.

**Comment by xsacha** [comment]
@biubug6 Isn't the entire idea of v2 and v3 that it increases performance (speed and accuracy)?
https://ai.googleblog.com/2018/04/mobilenetv2-next-generation-of-on.html
![image](https://user-images.githubusercontent.com/61218/65292205-d3b75f00-db99-11e9-84a7-370b2cfb6910.png)

**Comment by xsacha** [comment]
```
            elif net == 'shufflenet0.5':
            backbone = models.shufflenet_v2_x0_5(True)
            backbone.fc = nn.Sequential(
                nn.AdaptiveAvgPool2d((1,1)),
                View(),
                nn.Linear(256, backbone._stage_out_channels[0])
            )
            return_layers = {'stage2': 1, 'stage3': 2, 'stage4': 3}
```

Only downside is model is 24MB instead of 1.5MB :)
I'll train it out and tell you how it goes.

**Comment by biubug6** [comment]
If you are interested in this project, welcome to join us and pull requests. I also will do similar experiments and compare with you within the next two days.

**Comment by xsacha** [comment]
The shufflenet v2 0.5x isn't any faster when jitted. It is more accurate though.
Ideally I'd go for a shufflenet v2 0.25x but there isn't any pretrained.

**Comment by biubug6** [comment]
@xsacha shufflenet v2 0.5x is a little slow! The improvement of accuracy is not obvious.
The result is as follows(wider face val): 
original size: easy 90.75 mdium 88.30 hard 73.25
scale      size: easy 88.80 mdium 87.11 hard 80.77

**Comment by xsacha** [comment]
How much slower are you seeing? I'm getting same time with shufflenet v2 but better accuracy. Similar results to what you are getting.

**Comment by biubug6** [comment]
@xsacha  
using test.jpg and executing "detect.py":
net forward time: 0.0099
net forward time: 0.0104
net forward time: 0.0100
net forward time: 0.0103
net forward time: 0.0100
net forward time: 0.0104
net forward time: 0.0100
Mobilenet0.25 only consumes approximately 5~6ms.

**Comment by SnowRipple** [comment]
Hi guys!
I tried other version of mobilenet v1 (the only one I could find pretrained on Imagenet):
The pretrained model can be downloaded from here:
https://pan.baidu.com/s/1eRCxYKU

` def __init__(self):
        super(MobileNetV15, self).__init__()
        
        self.stage1 = nn.Sequential(
                conv_bn(3, 32, 2),    # 3
                conv_dw(32, 64, 1),   # 7
                conv_dw(64, 128, 2),  # 11
                conv_dw(128, 128, 1),  # 19
                conv_dw(128, 256, 2),  # 27
                conv_dw(256, 256, 1),  # 43
            )
        self.stage2 = nn.Sequential(
            conv_dw(256, 512, 2),
            conv_dw(512, 512, 1),
            conv_dw(512, 512, 1),
            conv_dw(512, 512, 1),
            conv_dw(512, 512, 1),
            conv_dw(512, 512, 1),
        )
        self.stage3 = nn.Sequential(
            conv_dw(512, 1024, 2),
            conv_dw(1024, 1024, 1),
        )
        self.avg =  nn.AvgPool2d(7)
        self.fc = nn.Linear(1024, 1000)`
The results are slightly better:

![mobilenets_comparison](https://user-images.githubusercontent.com/5250220/65429588-55afbe00-de0e-11e9-99dd-d5d2b9399a82.png)
I called it v2, but it is actually v1 but with more units. Still didn't finish comparing them.
@xsacha  would you mind sharing your shufflenet v2 net please?

**Comment by SnowRipple** [comment]
Interestingly, even though the model is much more "wide", the inference speed on test.jpg is still impressive:

net forward time: 0.0051
net forward time: 0.0052
net forward time: 0.0052
net forward time: 0.0052
net forward time: 0.0052
net forward time: 0.0052
net forward time: 0.0052
net forward time: 0.0051
net forward time: 0.0051
net forward time: 0.0052
net forward time: 0.0051
net forward time: 0.0051
net forward time: 0.0051

**Comment by biubug6** [comment]
You use mobilenetv1 instead of mobilenetv1X0.25. So it's going to be slow.

**Comment by SnowRipple** [comment]
But the net speeds (published above) for v1 are still fast (although model size jumped to 26MB). Net speeds are almost as good as 0.25

**Comment by xsacha** [comment]
If you have a decent GPU, the width will just occupy more of the GPU but will not affect latency.

The problem is when you run it on several at a time.

**Comment by SnowRipple** [comment]
@xsacha - makes sense!
I'm curious how it compares to your ShuffleNet, would you mind sparing me the google search for Imagenet pretrained version and sharing it please?

Also are you guys using some predefined evaluation script for widerface that generates these percentages for easy/medium/hard? I cannot see anything like this in this repository..

**Comment by biubug6** [comment]
Executing Evaluation widerface val steps in readme. It can get AP in easy/medium/hard.

**Comment by xsacha** [comment]
@SnowRipple https://drive.google.com/file/d/1rxWtlghq8Slj9IORguIBeyHcILrM4IR8/view?usp=sharing
I used torch.hub to get a pretrained shufflenet

```Python
    elif net == 'shufflenet':
            backbone = torch.hub.load('pytorch/vision', 'shufflenet_v2_x0_5', pretrained=True)
            backbone.eval()
            return_layers = {'stage2': 1, 'stage3': 2, 'stage4': 3}
            in_channels_stage2 = 24
```

**Comment by SnowRipple** [comment]
@xsacha mnasNet looks interesting and is provided in pytorch.hub

**Comment by xsacha** [comment]
mnasnet has the same issue as efficientnet. All the implementations consume too much memory.
I'm trying mnasnet_a1 with a batch size of 16 and will see how I go.

**Comment by 1093842024** [comment]
test test.jpg with detect.py with mobile0.25 in cpu mode, but the forward time is very slow(400ms), far slow than paper said,  has anyone know why?

**Comment by 1093842024** [comment]
test test.jpg with detect.py with mobile0.25 in cpu mode, but the forward time is very slow(400ms), far slow than paper said, does anyone know why?

**Comment by hegc** [comment]
@1093842024 I test the script on CPU(1 core), the inference time is about 70ms.

**Comment by rydenisbak** [comment]
@1093842024 may be yours cpu have no AVX2 instructions, or yours pytorch was compiled without avx2 support. Did you do warmup before time test?

**Comment by hegc** [comment]
So, any one uses the MobileNetV3 as backbone?

**Comment by xsacha** [comment]
It is very similar to v2 and has the same drawbacks.

**Comment by hegc** [comment]
@xsacha you mean the speed is lower than v1 on GPU?

**Comment by rydenisbak** [comment]
@hegc  I tried mv2 and mv3 but in my case mv2 was better

**Comment by hegc** [comment]
@rydenisbak Can you share the results of mobilev2 on the widerface?

**Comment by brealisty** [comment]
> @1093842024 I test the script on CPU(1 core), the inference time is about 70ms.

did u test the time of prior_box part, it's need more time than net forward.

**Comment by yy2yy** [comment]
Does anyone use MobileNetV3 as backbone get a better result?

**Comment by zdaiot** [comment]
@rydenisbak Can you share the results and speeds of mobilev2 on the widerface?

**Comment by quocnhat** [comment]
mobilenetv3 has layers channel of 16, 24, 40, 48 ... How can we return_layers? Because It requires a double size of layers (16 ->32 -> 64 ,...) any hind would be appreciated.

**Comment by quocnhat** [comment]
Here are some improvements: 
- mobile0.25: change output channel from 64-> 128 increases > 1 AP for each group of data. 
Easy Val AP: 0.9173621803802692
Medium Val AP: 0.8962414033919657
Hard Val AP: 0.7556337455094042
CPU 0.0959 GPU: 0.0044

- mobilenetv3 Small 128 output channel: 6.6M
Easy Val AP: 0.9322953274903119
Medium Val AP: 0.9116441066993648
Hard Val AP: 0.7799313057781694
CPU 0.0911 GPU 0.0080

- shufflenetv2 2.2M
Easy Val AP: 0.9165965195928714
Medium Val AP: 0.8900462910575522
Hard Val AP: 0.7416302903715322
CPU 0.1960 GPU 0.0153

**Comment by xsacha** [comment]
Mobilenetv3 looks good (for CPU)!
Thanks @quocnhat

**Comment by geoffzhang** [comment]
> Here are some improvements:
> 
> * mobile0.25: change output channel from 64-> 128 increases > 1 AP for each group of data.
>   Easy Val AP: 0.9173621803802692
>   Medium Val AP: 0.8962414033919657
>   Hard Val AP: 0.7556337455094042
>   CPU 0.0959 GPU: 0.0044
> * mobilenetv3 Small 128 output channel: 6.6M
>   Easy Val AP: 0.9322953274903119
>   Medium Val AP: 0.9116441066993648
>   Hard Val AP: 0.7799313057781694
>   CPU 0.0911 GPU 0.0080
> * shufflenetv2 2.2M
>   Easy Val AP: 0.9165965195928714
>   Medium Val AP: 0.8900462910575522
>   Hard Val AP: 0.7416302903715322
>   CPU 0.1960 GPU 0.0153

Hi, Can you share your mobilenet-v3 small?

**Comment by quocnhat** [comment]
https://github.com/rwightman/pytorch-image-models/tree/master/timm 
all the models are in this Github, Thanks #Rwightman 
btw, backbone mobilenetv2_100 seems very impresive ( fast and accurate : I.,e: 
easy|medium|hard|model_size|FPS (GPU)| FPS (CPU)
0.94127 | 0.9248 | 0.8284 | 8.3 | 167.67 | 4.545
GPU: GTX 1080, 
CPU: Intel(R) Core(TM) i5-7500 CPU @ 3.40GHz

**Comment by geoffzhang** [comment]
> CPU: Intel(R) Core(TM) i5-7500 CPU @ 3.40GHz
@quocnhat 
why is the mobilenetv2 slower than mobilenetv1 in my pc?

**Comment by bixiwen** [comment]
@quocnhat, would you mind sharing your mobilenetv2_100 net please? I don‘t find it in [https://github.com/rwightman/pytorch-image-models/tree/master/timm](url).

**Comment by quocnhat** [comment]
> @quocnhat, would you mind sharing your mobilenetv2_100 net please? I don‘t find it in [https://github.com/rwightman/pytorch-image-models/tree/master/timm](url).

https://rwightman.github.io/pytorch-image-models/ 

This is what you need, please check

**Comment by bixiwen** [comment]
> > @quocnhat, would you mind sharing your mobilenetv2_100 net please? I don‘t find it in [https://github.com/rwightman/pytorch-image-models/tree/master/timm](url).
> 
> https://rwightman.github.io/pytorch-image-models/
> 
> This is what you need, please check
Thank a lot

**Comment by bixiwen** [comment]
@quocnhat, thanks a lot

**Comment by bixiwen** [comment]
> mobilenetv3 has layers channel of 16, 24, 40, 48 ... How can we return_layers? Because It requires a double size of layers (16 ->32 -> 64 ,...) any hind would be appreciated.

@quocnhat, can you share how to solve this problem in mobilenetv2_100 net?

**Comment by quocnhat** [comment]
In file retinaface.py, replace in_channels_list to the three input FPN size of your model. for ex: 
in_channels_list = [ 16, 24, 40]

**Comment by noahzhy** [comment]
> Here are some improvements:
> 
> * mobile0.25: change output channel from 64-> 128 increases > 1 AP for each group of data.
>   Easy Val AP: 0.9173621803802692
>   Medium Val AP: 0.8962414033919657
>   Hard Val AP: 0.7556337455094042
>   CPU 0.0959 GPU: 0.0044
> * mobilenetv3 Small 128 output channel: 6.6M
>   Easy Val AP: 0.9322953274903119
>   Medium Val AP: 0.9116441066993648
>   Hard Val AP: 0.7799313057781694
>   CPU 0.0911 GPU 0.0080
> * shufflenetv2 2.2M
>   Easy Val AP: 0.9165965195928714
>   Medium Val AP: 0.8900462910575522
>   Hard Val AP: 0.7416302903715322
>   CPU 0.1960 GPU 0.0153

The CPU speed of shufflenetv2 in your testing result was slower than expected. Did you test the ShuffleNetV2-0.25x or ShuffleNet-0.5x(g=3) on it?

**Comment by quocnhat** [comment]
> > Here are some improvements:
> > 
> > * mobile0.25: change output channel from 64-> 128 increases > 1 AP for each group of data.
> >   Easy Val AP: 0.9173621803802692
> >   Medium Val AP: 0.8962414033919657
> >   Hard Val AP: 0.7556337455094042
> >   CPU 0.0959 GPU: 0.0044
> > * mobilenetv3 Small 128 output channel: 6.6M
> >   Easy Val AP: 0.9322953274903119
> >   Medium Val AP: 0.9116441066993648
> >   Hard Val AP: 0.7799313057781694
> >   CPU 0.0911 GPU 0.0080
> > * shufflenetv2 2.2M
> >   Easy Val AP: 0.9165965195928714
> >   Medium Val AP: 0.8900462910575522
> >   Hard Val AP: 0.7416302903715322
> >   CPU 0.1960 GPU 0.0153
> 
> The CPU speed of shufflenetv2 in your testing result was slower than expected. Did you test the ShuffleNetV2-0.25x or ShuffleNet-0.5x(g=3) on it?

If I did not mention anything else, It means that the network is the same as the default. pls check it

**Comment by luisfmnunes** [comment]
Anyone tested already with the recent ViT architectures in comparison with CNN based models?

**Comment by luisfmnunes** [comment]
> > @quocnhat, would you mind sharing your mobilenetv2_100 net please? I don‘t find it in [https://github.com/rwightman/pytorch-image-models/tree/master/timm](url).
> 
> https://rwightman.github.io/pytorch-image-models/
> 
> This is what you need, please check

Thank you for the reference, but I'm still having trouble to use timm models, like setting cfg parameters (specialy return_layers) and such. Could you please give more info on that? Thank you

Edit: I managed to use the parameter features_only to obtain the FPN, Thank you for the tips!
